### PR TITLE
Add docs for CA1859 analyzer rule

### PR DESCRIPTION
**Pull request by geeknoid**
## Summary

Describe your changes here.

Fixes #Issue_Number (if available)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/code-analysis/code-quality-rule-options.md](https://github.com/dotnet/docs/blob/66a108a753d557f19a7b908c22b363a44f853ed6/docs/fundamentals/code-analysis/code-quality-rule-options.md) | [Code quality rule configuration options](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/code-quality-rule-options?branch=pr-en-us-35182) |
| [docs/fundamentals/code-analysis/quality-rules/ca1859.md](https://github.com/dotnet/docs/blob/66a108a753d557f19a7b908c22b363a44f853ed6/docs/fundamentals/code-analysis/quality-rules/ca1859.md) | [docs/fundamentals/code-analysis/quality-rules/ca1859](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1859?branch=pr-en-us-35182) |
| [docs/fundamentals/code-analysis/quality-rules/index.md](https://github.com/dotnet/docs/blob/66a108a753d557f19a7b908c22b363a44f853ed6/docs/fundamentals/code-analysis/quality-rules/index.md) | [Code quality rules](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/index?branch=pr-en-us-35182) |
| [docs/fundamentals/code-analysis/quality-rules/performance-warnings.md](https://github.com/dotnet/docs/blob/66a108a753d557f19a7b908c22b363a44f853ed6/docs/fundamentals/code-analysis/quality-rules/performance-warnings.md) | [Performance rules](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/performance-warnings?branch=pr-en-us-35182) |

### CODE REVIEW

**Review by Youssef1313 — CHANGES_REQUESTED** [review]
This is missing a toc entry I believe

File: docs/fundamentals/code-analysis/quality-rules/performance-warnings.md [hunk 1]
```diff
@@ -67,4 +67,5 @@ Performance rules support high-performance libraries and applications.
 | [CA1854: Prefer the 'IDictionary.TryGetValue(TKey, out TValue)' method](ca1854.md) | Prefer 'TryGetValue' over a Dictionary indexer access guarded by a 'ContainsKey' check. 'ContainsKey' and the indexer both look up the key, so using 'TryGetValue' avoids the extra lookup. |
 | [CA1855: Use Span\<T>.Clear() instead of Span\<T>.Fill()](ca1855.md) | It's more efficient to call <xref:System.Span%601.Clear?displayProperty=nameWithType> than to call <xref:System.Span%601.Fill(%600)?displayProperty=nameWithType> to fill the elements of the span with a default value. |
 | [CA1858: Use StartsWith instead of IndexOf](ca1858.md) | It's more efficient to call <xref:System.String.StartsWith%2A?displayProperty=nameWithType> than to call <xref:System.String.IndexOf%2A?displayProperty=nameWithType> to check whether a string starts with a given prefix. |
+| [CA1859: Use concrete types when possible for improved performance](ca1859.md) | Code is using interface types or abstract types within its implementation, leading to unecessary interface calls or virtual calls. |
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/performance-warnings.md:70]:
```suggestion
| [CA1859: Use concrete types when possible for improved performance](ca1859.md) | Code uses interface types or abstract types, leading to unnecessary interface calls or virtual calls. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1859.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+{
+    public virtual void M() { }
+}
+
+sealed class B
+{ }
+
+internal class C
+{
+    private readonly B _a = new B();
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1859.md:82]:
```suggestion
    private readonly B _b = new B();
```

File: docs/fundamentals/code-analysis/quality-rules/ca1859.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    public void Trigger()
+    {
+        // this performs a virtual call because _a is defined as an abstract class
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1859.md:86]:
```suggestion
```

File: docs/fundamentals/code-analysis/quality-rules/ca1859.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        _a.M();
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1859.md:87]:
```suggestion
        _b.M();
```

File: docs/fundamentals/code-analysis/quality-rules/ca1859.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+| **Enabled by default in .NET 8**    | No                                     |
+
+## Cause
+
+Code is using interface types or abstract types within its implementation, leading to unecessary
+interface calls or virtual calls.
+
+## Rule description
+
+This rule recommends upgrading the type of individual local variables, fields, properties, method parameters, and method return types from interface or abstract
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1859.md:33]:
```suggestion
This rule recommends upgrading the type of specific local variables, fields, properties, method parameters, and method return types from interface or abstract
```

File: docs/fundamentals/code-analysis/code-quality-rule-options.md [hunk 6]
```diff
@@ -73,23 +73,31 @@ The syntax for configuring an option for a *specific* rule is as follows:
...
-- [output_kind](#output_kind)
-- [required_modifiers](#required_modifiers)
-- [exclude_extension_method_this_parameter](#exclude_extension_method_this_parameter)
-- [null_check_validation_methods](#null_check_validation_methods)
-- [additional_string_formatting_methods](#additional_string_formatting_methods)
-- [excluded_type_names_with_derived_types](#excluded_type_names_with_derived_types)
-- [excluded_symbol_names](#excluded_symbol_names)
-- [disallowed_symbol_names](#disallowed_symbol_names)
+- [Code quality rule configuration options](#code-quality-rule-configuration-options)
+  - [Option scopes](#option-scopes)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/code-quality-rule-options.md:77]: These aren't in this `## Options` section. Did you use a tool to generate this list?

**Review by gewarren — COMMENTED** [review]
The TOC to update is at https://github.com/dotnet/docs/blob/main/docs/navigate/tools-diagnostics/toc.yml.

Thanks for adding this!

File: docs/fundamentals/code-analysis/code-quality-rule-options.md (hunk 6 above)
Comment (geeknoid) [reply docs/fundamentals/code-analysis/code-quality-rule-options.md:77]: I have no idea where this stuff comes from. I'll undo those changes.

Comment (geeknoid) [reply docs/fundamentals/code-analysis/code-quality-rule-options.md:77]: Turns out this is from VS Code. If I edit anything in this file, it insists on replacing this set of lines when I save. Very strange.

**Review by gewarren — APPROVED** [review]

**Pull request merged**